### DISCUSSION TITLE
Feature request: 2MB ESP32 build

### DISCUSSION THREAD

**Discussion by Tediore** [arendst/Tasmota#12673] [Ideas]
The ESP32-C3F comes in two variants: one with 4MB embedded flash and one with 2MB external flash. I accidentally bought the one with 2MB flash and as a result, I can't run Tasmota on it. Would be great to have a 2MB ESP32 build for those of us unfortunate enough to buy that variant instead of the 4MB. 

@Jason2866 I think you might have some insight on this?

**Comment by s-hadinger** [comment]
I don't see why it wouldn't flash. Maybe you need to remove ora_1 partition from the partition csv. 

But keep in mind that you won't be able to OTA flash. Only serial.

**Comment by Jason2866** [comment]
You can try https://github.com/Jason2866/Tasmota-specials/tree/firmware
For flashing look in folder https://github.com/Jason2866/Tasmota-specials/tree/firmware/firmware/tasmota32/ESP32-C3_needed_files
The only difference is the partitions.bin!

**Reply by effelle** [reply]
Nice one J!

**Reply by Tediore** [reply]
Thank you! I'll give it a try.

**Reply by Tediore** [reply]
Finally had an opportunity to try. Works great, thank you @Jason2866 !

**Reply by Jason2866** [reply]
You can update via serial (when you use the 2M partions.bin!) the 4M ESP32-C3 firmware build.
The firmware itself does NOT differ. It is only the partition.bin

**Reply by Tediore** [reply]
Perfect, thank you for clarifying.

**Reply by EdBrisebois** [reply]
Same issue as Tedoire, vendor said 4M but delivered 2M.  The web interface appears to load the C3 2M bluetooth FW but it just bootloops. Serial spits out:
15:54:56.693 -> 52-g209a0d389 2nd stage bootloader
15:54:56.693 -> I (53) boot: compile time 09ESP-ROM:esp32c3-20200918
15:54:56.693 -> Build:Sep 18 2020
15:54:56.693 -> rst:0x1 (POWERON),boot:0xc (SPI_FAST_FLASH_BOOT)
15:54:56.693 -> SPIWP:0xee
15:54:56.693 -> mode:DIO, clock div:2
15:54:56.693 -> load:0x3fcd6100,len:0x108c
15:54:56.693 -> load:0x403ce000,len:0x860
15:54:56.693 -> load:0x403d0000,len:0x27c4
15:54:56.693 -> SHA-256 comparison failed:
15:54:56.693 -> Calculated: 7b8abcd13b9167c3b92e62074b140340e9ed1bb2c4ca06ce9a4ad60554217415
15:54:56.693 -> Expected: 4b8e7048808df2cfbf3670209bf7b63f662976dda1c12d75e112aaa36d43de4d
15:54:56.693 -> Attempting to boot anyway...
15:54:56.693 -> entry 0x403ce000
15:54:56.693 -> I (50) boot: ESP-IDF v4.4-dev-1852-g209a0d389 2nd stage bootloader
15:54:56.693 -> I (50) boot: compile time 09:24:01
15:54:56.717 -> I (51) boot: chip revision: 2
15:54:56.717 -> I (51) boot_comm: chip revision: 2, min. bootloader chip revision: 0
15:54:56.717 -> I (57) boot.esp32c3: SPI Speed      : 40MHz
15:54:56.717 -> I (61) boot.esp32c3: SPI Mode       : DIO
15:54:56.717 -> I (65) boot.esp32c3: SPI Flash Size : 2MB
15:54:56.717 -> I (68) boot: Enabling RNG early entropy source...
15:54:56.717 -> E (73) flash_parts: partition 3 invalid - offset 0x1e0000 size 0x1d0000 exceeds flash chip size 0x200000
15:54:56.717 -> E (82) boot: Failed to verify partition table
15:54:56.717 -> E (86) boot: load partition table error!
15:54:56.752 -> ESP-ROM:esp32c3-20200918
15:54:56.752 -> Build:Sep 18 2020
15:54:56.752 -> rst:0x3 (RTC_SW_SYS_RST),boot:0xc (SPI_FAST_FLASH_BOOT)
15:54:56.752 -> Saved PC:0x403d1188
15:54:56.752 -> SPIWP:0xee
15:54:56.752 -> mode:DIO, clock div:2
15:54:56.752 -> load:0x3fcd6100,len:0x108c
15:54:56.752 -> load:0x403ce000,len:0x860
15:54:56.752 -> load:0x403d0000,len:0x27c4
15:54:56.788 -> SHA-256 comparison failed:
15:54:56.788 -> Calculated: 7b8abcd13b9167c3b92e62074b140340e9ed1bb2c4ca06ce9a4ad60554217415
15:54:56.788 -> Expected: 4b8e7048808df2cfbf3670209bf7b63f662976dda1c12d75e112aaa36d43de4d
15:54:56.788 -> Attempting to boot anyway...
15:54:56.788 -> entry 0x403ce000

I downloaded from the specials but the second link shows 404.   
Any Ideas?

**Reply by sfromis** [reply]
Current links:
https://github.com/tasmota/install/tree/main/firmware/unofficial
https://github.com/tasmota/install/tree/main/static/esp32-c3

**Comment by EdBrisebois** [comment]
That was one hail mary for OSX!
It finally did something, at least it is not boot looping now.
No sign of a web server as stated.
Verbose:
17:20:49.889 -> ESP-ROM:esp32c3-20200918
17:20:49.889 -> Build:Sep 18 2020
17:20:49.889 -> rst:0x1 (POWERON),boot:0xc (SPI_FAST_FLASH_BOOT)
17:20:49.889 -> SPIWP:0xee
17:20:49.889 -> mode:DIO, clock div:2
17:20:49.889 -> load:0x3fcd6100,len:0x108c
17:20:49.889 -> load:0x403ce000,len:0x860
17:20:49.889 -> load:0x403d0000,len:0x27c4
17:20:49.889 -> SHA-256 comparison failed:
17:20:49.889 -> Calculated: 33fdca06eef9863dc8c7198bbb94d3156188f4484a809599a3fc444eea32ec61
17:20:49.889 -> Expected: 4b8e7048808df2cfbf3670209bf7b63f662976dda1c12d75e112aaa36d43de4d
17:20:49.927 -> Attempting to boot anyway...
17:20:49.927 -> entry 0x403ce000
17:20:49.927 -> I (50) boot: ESP-IDF v4.4-dev-1852-g209a0d389 2nd stage bootloader
17:20:49.927 -> I (50) boot: compile time 09:24:01
17:20:49.927 -> I (51) boot: chip revision: 2
17:20:49.927 -> I (51) boot_comm: chip revision: 2, min. bootloader chip revision: 0
17:20:49.927 -> I (57) boot.esp32c3: SPI Speed      : 40MHz
17:20:49.927 -> I (61) boot.esp32c3: SPI Mode       : DIO
17:20:49.927 -> I (65) boot.esp32c3: SPI Flash Size : 4MB
17:20:49.927 -> I (68) boot: Enabling RNG early entropy source...
17:20:49.963 -> I (73) boot: Partition Table:
17:20:49.963 -> I (75) boot: ## Label            Usage          Type ST Offset   Length
17:20:49.963 -> I (82) boot:  0 nvs              WiFi data        01 02 00009000 00005000
17:20:49.963 -> I (88) boot:  1 otadata          OTA data         01 00 0000e000 00002000
17:20:49.963 -> I (95) boot:  2 app0             OTA app          00 10 00010000 001d0000
17:20:49.963 -> I (101) boot:  3 app1             OTA app          00 11 001e0000 001d0000
17:20:50.001 -> I (108) boot:  4 spiffs           Unknown data     01 82 003b0000 00050000
17:20:50.001 -> I (114) boot: End of partition table
17:20:50.001 -> I (118) boot_comm: chip revision: 2, min. application chip revision: 0
17:20:50.001 -> I (124) esp_image: segment 0: paddr=00010020 vaddr=3c110020 size=2a2e0h (172768) map
17:20:50.039 -> I (168) esp_image: segment 1: paddr=0003a308 vaddr=3fc8d400 size=03228h ( 12840) load
17:20:50.039 -> I (171) esp_image: segment 2: paddr=0003d538 vaddr=40380000 size=02ae0h ( 10976) load
17:20:50.039 -> I (174) esp_image: segment 3: paddr=00040020 vaddr=42000020 size=10fdc0h (1113536) map
17:20:50.295 -> I (412) esp_image: segment 4: paddr=0014fde8 vaddr=40382ae0 size=0a798h ( 42904) load
17:20:50.295 -> I (423) esp_image: segment 5: paddr=0015a588 vaddr=50000000 size=00004h (     4) load
17:20:50.295 -> I (423) esp_image: segment 6: paddr=0015a594 vaddr=50000008 size=00010h (    16) load
17:20:50.295 -> I (431) boot: Loaded app from partition at offset 0x10000
17:20:50.295 -> I (432) boot: Disabling RNG early entropy source...
17:20:50.331 -> 
17:20:50.331 -> 00:00:00.002-248/75 HDW: ESP32-C3 
17:20:50.331 -> 00:00:00.003-248/75 CFG: Use defaults
17:20:50.368 -> 00:00:00.022 QPC: Reset
17:20:50.368 -> 00:00:00.045 BRY: Berry initialized, RAM used=3758
17:20:50.368 -> 00:00:00.048 BRY: no 'preinit.be'
17:20:50.368 -> 00:00:00.054 Project tasmota - Tasmota Version 10.0.0.2(tasmota)-2_0_1(2021-11-10T20:29:26)
17:20:50.368 -> 00:00:00.058 BRY: no 'autoexec.be'
17:20:50.691 -> 00:00:00.358 WIF: WifiManager active for 3 minutes
17:20:50.799 -> 00:00:00.461 WIF: Retry set Mode...
17:20:50.871 -> 00:00:00.562 WIF: Retry set Mode...
17:20:51.535 -> 00:00:01.175 HTP: Web server active on tasmota-862BE4-3044 with IP address 0.0.0.0
17:20:56.888 -> 00:00:06.524 QPC: Reset

**Reply by EdBrisebois** [reply]
I tried going back to the webinstaller hoping the partitions are happy. It loads something pointing to a C3_2M variant but it just boot loops again.  I reinstalled with the command line again and it is back to the last posted verbose.  However, does not seem to be barking (noRF).  I compared to the working 4M module, I see lots of barking.  Either the wifi is not ON or a faulty 2M module :o\

**Reply by Jason2866** [reply]
This `17:20:51.535 -> 00:00:01.175 HTP: Web server active on tasmota-862BE4-3044 with IP address 0.0.0.0` is the info for not working wifi. IP address is not valid.

**Reply by Jason2866** [reply]
Just tried. Works for me
```
00:00:00.002 HDW: ESP32-C3 
00:00:00.015 UFS: FlashFS mounted with 272 kB free
00:00:00.030 CFG: Loaded from File, Count 13633
00:00:00.034 CFG: CR 472/699, Busy 0
00:00:00.038 CFG: CR 472/699, Busy 0
00:00:00.039 ROT: Mode 1
00:00:00.040 BRY: GC from 1242 to 1242 bytes (in 0 ms)
00:00:00.042 BRY: Stack resized from 160 to 328 bytes
00:00:00.059 BRY: GC from 3590 to 3012 bytes (in 0 ms)
00:00:00.082 BRY: Stack resized from 328 to 504 bytes
00:00:00.090 BRY: GC from 4577 to 3968 bytes (in 1 ms)
00:00:00.091 BRY: Berry initialized, RAM used=3968
00:00:00.098 BRY: no 'preinit.be'
00:00:00.101 SRC: Restart
00:00:00.102 Project tasmota - Tasmota Version 10.0.0.2(tasmota)-2_0_1(2021-11-10T20:29:43)
00:00:00.104 M32: init: request callbacks
00:00:00.122 BRY: no 'autoexec.be'
00:00:00.378 WIF: Attempting connection...
00:00:00.761 WIF: Connecting to AP2 Jason_Home_WLAN Channel 9 BSSId 00:A0:57:2A:BD:19 in mode 11n as sonoff-3B0814-2068...
00:00:01.592 WIF: Attempting connection...
00:00:02.593 WIF: Attempting connection...
00:00:03.738 WIF: Attempting connection...
00:00:04.843 WIF: Attempting connection...
00:00:05.843 WIF: Connected
00:00:05.844 BLE: Put WiFi modem in sleep mode
00:00:06.098 HTP: Web server active on sonoff-3B0814-2068 with IP address 192.168.2.160
00:00:09.188 RTC: UTC 2021-11-11T10:50:03, DST 2021-03-28T02:00:00, STD 2021-10-31T03:00:00
11:50:03.002 BLE: MasterEnable->1
11:50:03.003 BLE: BLETask:Starting NimBLE
11:50:03.297 APP: Boot Count 11
11:50:03.516 MQT: Attempting connection...
11:50:03.588 MQT: Connected
11:50:03.593 MQT: tele/sonoff-3B0814/LWT = Online (retained)
```

**Comment by EdBrisebois** [comment]
Probably why I don't see an AP in my wifi list.
Why is the firmware setting the IP to 0.0.0.0?

I used the files from:
https://github.com/tasmota/install/blob/main/firmware/unofficial/tasmota32c3_2M.bin
https://github.com/tasmota/install/tree/main/static/esp32-c3

command line:
/Users/***********/Library/Arduino15/packages/esp32/tools/esptool_py/3.0.0/esptool --chip esp32c3 --port /dev/cu.wchusbserial1440 --baud 115200 --before default_reset --after hard_reset write_flash -z --flash_mode dout --flash_freq 40m --flash_size detect 0x0000 /Users/***********/bootloader_dout_40m.bin 0x8000 partitions_2M.bin 0xe000 boot_app0.bin 0x10000 tasmota32c3_2M.bin

After loading, still comes back with:  -> 00:00:01.175 HTP: Web server active on tasmota-862BE4-3044 with IP address 0.0.0.0

In the executing string from the command line:
Configuring flash size...
Warning: Could not auto-detect Flash size (FlashID=0x0, SizeID=0x0), defaulting to 4MB
Flash params set to 0x0320

Not writing to the 2M flash correctly?

________________________________

**Reply by Jason2866** [reply]
Important you have to use the 2M partition file! https://github.com/Jason2866/Tasmota-specials/raw/firmware/static/esp32-c3/partitions_2M.bin
And use ONLY esptool.py v.3.1 or newer for flashing a C3
Better to use esptool.py v.3.2 since older versions do not detect all C3 correctly!
https://github.com/tasmota/esptool/releases/download/v3.2/esptool-v3.2.zip

**Reply by sfromis** [reply]
Your command looks "risky" with the bootloader path having a source directory specified, and the other files not. Results depends on your current directory. Do check the messages from esptool.py carefully....

**Reply by EdBrisebois** [reply]
I did use the 2M parition file.

I'll check the esptool.

I put the bin files in the specified path, so it does pulls them out.

**Comment by Jason2866** [comment]
Tried the Tasmota WebInstaller. For me it just works.

**Reply by EdBrisebois** [reply]
It just boots loops when I load that way.  Tried both the C3-2M with and with out bluetooth.

**Reply by Jason2866** [reply]
Esptool.py v.3.0 (Arduino version) does NOT support the C3 correctly. You will probably never get a working successfull flash
`Warning: Could not auto-detect Flash size (FlashID=0x0, SizeID=0x0), defaulting to 4MB`

**Reply by Jason2866** [reply]
Where did you get the board? It is a old early revision! 
`17:20:50.001 -> I (118) boot_comm: chip revision: 2, min. application chip revision: 0`
chip revision should be: 3

**Reply by EdBrisebois** [reply]
Ok, I surrender, wasted over 2 days on this $$.
Thanks for the help guys, cheers:o)

**Comment by EdBrisebois** [comment]
I bought the modules and put them on my 8266-12F boards.
4M worked no problem.
2M is still kicking my a$$.

unistalled 3.0...VPN was blocking my install of 3.2... now 3.2 installed but not the folder where 3.0 was...still looking
________________________________
